# MRETT Results

Complete spreadsheet: https://docs.google.com/spreadsheets/d/11uk0WpQzjVBOWs99cev9buajG1PalFdC5mtdynhqejU/edit?fbclid=IwAR2KA8_pks7JY4Q_P7XLrNSawnmysBJo2FnAFvjUe5r7XiAlAi134Z-PXNw#gid=1384530880 

MRETT21: 1gE2gzncRuiwzY4xgKEbVDjSBUde8LGc7u9pIMQ0C6bQ 

https://rgtdb.com/events/94944

## Importing Google Spreadsheet into Pandas

https://stackoverflow.com/questions/19611729/getting-google-spreadsheet-csv-into-a-pandas-dataframe

```python
pd.read_csv('https://docs.google.com/spreadsheets/d/' + 
                   '0Ak1ecr7i0wotdGJmTURJRnZLYlV3M2daNTRubTdwTXc' +
                   '/export?gid=0&format=csv',
                   # Set first column as rownames in data frame
                   index_col=0,
                   # Parse column values to datetime
                   parse_dates=['Quradate']
                  )
```

## APIs

https://docs.google.com/spreadsheets/d/11uk0WpQzjVBOWs99cev9buajG1PalFdC5mtdynhqejU/values/MRETT22

https://docs.google.com/spreadsheets/d/11uk0WpQzjVBOWs99cev9buajG1PalFdC5mtdynhqejU/edit#gid=171816323&range=C17

GC: https://docs.google.com/spreadsheets/d/11uk0WpQzjVBOWs99cev9buajG1PalFdC5mtdynhqejU/edit?fbclid=IwAR2KA8_pks7JY4Q_P7XLrNSawnmysBJo2FnAFvjUe5r7XiAlAi134Z-PXNw#gid=1384530880
21: https://docs.google.com/spreadsheets/d/11uk0WpQzjVBOWs99cev9buajG1PalFdC5mtdynhqejU/edit?fbclid=IwAR2KA8_pks7JY4Q_P7XLrNSawnmysBJo2FnAFvjUe5r7XiAlAi134Z-PXNw#gid=0
22: https://docs.google.com/spreadsheets/d/11uk0WpQzjVBOWs99cev9buajG1PalFdC5mtdynhqejU/edit?fbclid=IwAR2KA8_pks7JY4Q_P7XLrNSawnmysBJo2FnAFvjUe5r7XiAlAi134Z-PXNw#gid=601708178
23: https://docs.google.com/spreadsheets/d/11uk0WpQzjVBOWs99cev9buajG1PalFdC5mtdynhqejU/edit?fbclid=IwAR2KA8_pks7JY4Q_P7XLrNSawnmysBJo2FnAFvjUe5r7XiAlAi134Z-PXNw#gid=171816323
24: https://docs.google.com/spreadsheets/d/11uk0WpQzjVBOWs99cev9buajG1PalFdC5mtdynhqejU/edit?fbclid=IwAR2KA8_pks7JY4Q_P7XLrNSawnmysBJo2FnAFvjUe5r7XiAlAi134Z-PXNw#gid=1769801856
25: https://docs.google.com/spreadsheets/d/11uk0WpQzjVBOWs99cev9buajG1PalFdC5mtdynhqejU/edit?fbclid=IwAR2KA8_pks7JY4Q_P7XLrNSawnmysBJo2FnAFvjUe5r7XiAlAi134Z-PXNw#gid=1953569581
26: https://docs.google.com/spreadsheets/d/11uk0WpQzjVBOWs99cev9buajG1PalFdC5mtdynhqejU/edit?fbclid=IwAR2KA8_pks7JY4Q_P7XLrNSawnmysBJo2FnAFvjUe5r7XiAlAi134Z-PXNw#gid=1178762166

In [25]:
import pandas as pd

df_MRETT_GC = pd.read_csv('https://docs.google.com/spreadsheets/d/' + 
                   '11uk0WpQzjVBOWs99cev9buajG1PalFdC5mtdynhqejU' +
                   '/export?gid=' + '1384530880' + '&format=csv',
                   # Set first column as rownames in data frame
                   index_col=0,
                   # Parse column values to datetime
                   #parse_dates=['Quradate']
                  )

#df_MRETT22 = pd.read_csv('https://docs.google.com/spreadsheets/d/' + '1gIz_Gs6uxUrMUJKk7SrflaAiKeOKScQfvNqnX4bADV0' + '/export?gid=' + '1667031214' + '&format=csv', index_col=0)

df_MRETT_GC


Team  Total MRETT21 MRETT22 MRETT23 MRETT24 MRETT25  \
#                                                                          
1                  eCKD 1   1200     200     200     200     200     200   
2            The Pedalers   1110     170     190     190     190     180   
3            Rasio Racing   1100     190     180     180     180     190   
4              Pocomotion   1020     175     165     170     165     175   
5               OTR Black   1010     165     160     175     175     160   
6             Moon Riders    956     160     141     160     160     170   
7                OTR Blue    920     155     170     145     145     150   
8               OTR Green    908     141     137     165     150     155   
9   Team Lou Racing Squad    882     137     145     150     155     145   
10                 eCKD 2    834     145     133     137     137     141   
11     KISS Racing Team 1    815     150     175     155     170     165   
12    Westerley CC Purple    814     129     129     133     141     137   
13          WKG Renegades    520     125     125     141     129           
14             Westerlies    504             121     129     125     129   
15           Team Lou Too    403     NaN     NaN     NaN     133     133   
16           RGT France 1    335     180     155                           
17              Les Watts    150             150                           
18        Tricky Allsorts    133     133                                   

   MRETT26  
#           
1      200  
2      190  
3      180  
4      170  
5      175  
6      165  
7      155  
8      160  
9      150  
10     141  
11          
12     145  
13          
14          
15     137  
16          
17          
18

## Wait for Spreadsheet to Initialize

Since the spreadsheet reads data from other sheets, it may take several seconds for it to properly fetch and calculate all values.

In [26]:
import time

for i in range(1):
    time.sleep(5) 
    pd.read_csv('https://docs.google.com/spreadsheets/d/' + 
                    '11uk0WpQzjVBOWs99cev9buajG1PalFdC5mtdynhqejU' +
                    '/export?' + '' + '&format=csv',
                    index_col=1,
                    )

## Get all Sheet Tabs, Including GC

In [27]:
mrett_gids = {
    'GC' : '1384530880',
    '21' : '0',
    '22' : '601708178',
    '23' : '171816323',
    '24' : '1769801856',
    '25' : '1953569581',
    '26' : '1178762166'
}

mrett_finished = {
    'GC' : False,
    '21' : True,
    '22' : True,
    '23' : True,
    '24' : True,
    '25' : True,
    '26' : False
}

df_mretts = {}

for mrett in mrett_gids.keys():
    print(mrett, mrett_gids[mrett])
    gid = mrett_gids[mrett]
    if mrett_finished[mrett] != True:
        print('Updating from spreadsheet')
        df_mretts[mrett] = pd.read_csv('https://docs.google.com/spreadsheets/d/' + '11uk0WpQzjVBOWs99cev9buajG1PalFdC5mtdynhqejU' + '/export?gid=' + gid + '&format=csv', index_col=1, na_values=['', ' '])
    else:
        print('Using stored values')
        filename = mrett + '.pickle'
        df_mretts[mrett] = pd.read_pickle(filename)

df_mretts['GC']

GC 1384530880
Updating from spreadsheet
21 0
Using stored values
22 601708178
Using stored values
23 171816323
Using stored values
24 1769801856
Using stored values
25 1953569581
Using stored values
26 1178762166
Updating from spreadsheet


#  Total  MRETT21  MRETT22  MRETT23  MRETT24  MRETT25  \
Team                                                                            
eCKD 1                  1   1200    200.0    200.0    200.0    200.0    200.0   
The Pedalers            2   1110    170.0    190.0    190.0    190.0    180.0   
Rasio Racing            3   1100    190.0    180.0    180.0    180.0    190.0   
Pocomotion              4   1020    175.0    165.0    170.0    165.0    175.0   
OTR Black               5   1010    165.0    160.0    175.0    175.0    160.0   
Moon Riders             6    956    160.0    141.0    160.0    160.0    170.0   
OTR Blue                7    920    155.0    170.0    145.0    145.0    150.0   
OTR Green               8    908    141.0    137.0    165.0    150.0    155.0   
Team Lou Racing Squad   9    882    137.0    145.0    150.0    155.0    145.0   
eCKD 2                 10    834    145.0    133.0    137.0    137.0    141.0   
KISS Racing Team 1     11    815    150.0    175.0    155.0    170.0    165.0   
Westerley CC Purple    12    814    129.0    129.0    133.0    141.0    137.0   
WKG Renegades          13    520    125.0    125.0    141.0    129.0      NaN   
Westerlies             14    504      NaN    121.0    129.0    125.0    129.0   
Team Lou Too           15    403      NaN      NaN      NaN    133.0    133.0   
RGT France 1           16    335    180.0    155.0      NaN      NaN      NaN   
Les Watts              17    150      NaN    150.0      NaN      NaN      NaN   
Tricky Allsorts        18    133    133.0      NaN      NaN      NaN      NaN   

                       MRETT26  
Team                            
eCKD 1                   200.0  
The Pedalers             190.0  
Rasio Racing             180.0  
Pocomotion               170.0  
OTR Black                175.0  
Moon Riders              165.0  
OTR Blue                 155.0  
OTR Green                160.0  
Team Lou Racing Squad    150.0  
eCKD 2                   141.0  
KISS Racing Team 1         NaN  
Westerley CC Purple      145.0  
WKG Renegades              NaN  
Westerlies                 NaN  
Team Lou Too             137.0  
RGT France 1               NaN  
Les Watts                  NaN  
Tricky Allsorts            NaN

In [28]:
if pd.isna(df_mretts['GC']['MRETT21'].iloc[0]):
    df_mretts['GC'].at['eCKD 1', 'MRETT21'] = 200
    df_mretts['GC'].at['Rasio Racing', 'MRETT21'] = 190
    df_mretts['GC'].at['The Pedalers', 'MRETT21'] = 170
    df_mretts['GC'].at['Pocomotion', 'MRETT21'] = 175
    df_mretts['GC'].at['OTR Black', 'MRETT21'] = 165
    df_mretts['GC'].at['KISS Racing Team 1', 'MRETT21'] = 150
    df_mretts['GC'].at['Moon Riders', 'MRETT21'] = 160
    df_mretts['GC'].at['OTR Blue', 'MRETT21'] = 155
    df_mretts['GC'].at['OTR Green', 'MRETT21'] = 141
    df_mretts['GC'].at['Team Lou Racing Squad', 'MRETT21'] = 137
    df_mretts['GC'].at['eCKD 2', 'MRETT21'] = 145
    df_mretts['GC'].at['Westerley CC Purple', 'MRETT21'] = 129
    df_mretts['GC'].at['WKG Renegades', 'MRETT21'] = 125
    df_mretts['GC'].at['Westerlies', 'MRETT21'] = 0
    df_mretts['GC'].at['RGT France 1', 'MRETT21'] = 189
    df_mretts['GC'].at['Team Lou Too', 'MRETT21'] = 0
    df_mretts['GC'].at['Les Watts', 'MRETT21'] = 0
    df_mretts['GC'].at['Tricky Allsorts', 'MRETT21'] = 133
    df_mretts['GC']['Total']  = df_mretts['GC']['Total'] + df_mretts['GC']['MRETT21'] 


df_mretts['GC']

#  Total  MRETT21  MRETT22  MRETT23  MRETT24  MRETT25  \
Team                                                                            
eCKD 1                  1   1200    200.0    200.0    200.0    200.0    200.0   
The Pedalers            2   1110    170.0    190.0    190.0    190.0    180.0   
Rasio Racing            3   1100    190.0    180.0    180.0    180.0    190.0   
Pocomotion              4   1020    175.0    165.0    170.0    165.0    175.0   
OTR Black               5   1010    165.0    160.0    175.0    175.0    160.0   
Moon Riders             6    956    160.0    141.0    160.0    160.0    170.0   
OTR Blue                7    920    155.0    170.0    145.0    145.0    150.0   
OTR Green               8    908    141.0    137.0    165.0    150.0    155.0   
Team Lou Racing Squad   9    882    137.0    145.0    150.0    155.0    145.0   
eCKD 2                 10    834    145.0    133.0    137.0    137.0    141.0   
KISS Racing Team 1     11    815    150.0    175.0    155.0    170.0    165.0   
Westerley CC Purple    12    814    129.0    129.0    133.0    141.0    137.0   
WKG Renegades          13    520    125.0    125.0    141.0    129.0      NaN   
Westerlies             14    504      NaN    121.0    129.0    125.0    129.0   
Team Lou Too           15    403      NaN      NaN      NaN    133.0    133.0   
RGT France 1           16    335    180.0    155.0      NaN      NaN      NaN   
Les Watts              17    150      NaN    150.0      NaN      NaN      NaN   
Tricky Allsorts        18    133    133.0      NaN      NaN      NaN      NaN   

                       MRETT26  
Team                            
eCKD 1                   200.0  
The Pedalers             190.0  
Rasio Racing             180.0  
Pocomotion               170.0  
OTR Black                175.0  
Moon Riders              165.0  
OTR Blue                 155.0  
OTR Green                160.0  
Team Lou Racing Squad    150.0  
eCKD 2                   141.0  
KISS Racing Team 1         NaN  
Westerley CC Purple      145.0  
WKG Renegades              NaN  
Westerlies                 NaN  
Team Lou Too             137.0  
RGT France 1               NaN  
Les Watts                  NaN  
Tricky Allsorts            NaN

In [29]:

""" df_mretts['GC'].at['eCKD 1', 'MRETT21'] = 200
df_mretts['GC'].at['Rasio Racing', 'MRETT21'] = 190
df_mretts['GC'].at['The Pedalers', 'MRETT21'] = 170
df_mretts['GC'].at['Pocomotion', 'MRETT21'] = 175
df_mretts['GC'].at['OTR Black', 'MRETT21'] = 165
df_mretts['GC'].at['KISS Racing Team 1', 'MRETT21'] = 150
df_mretts['GC'].at['Moon Riders', 'MRETT21'] = 160
df_mretts['GC'].at['OTR Blue', 'MRETT21'] = 155
df_mretts['GC'].at['OTR Green', 'MRETT21'] = 141
df_mretts['GC'].at['Team Lou Racing Squad', 'MRETT21'] = 137
df_mretts['GC'].at['eCKD 2', 'MRETT21'] = 145
df_mretts['GC'].at['Westerley CC Purple', 'MRETT21'] = 129
df_mretts['GC'].at['WKG Renegades', 'MRETT21'] = 125
df_mretts['GC'].at['Westerlies', 'MRETT21'] = 0
df_mretts['GC'].at['RGT France 1', 'MRETT21'] = 189
df_mretts['GC'].at['Team Lou Too', 'MRETT21'] = 0
df_mretts['GC'].at['Les Watts', 'MRETT21'] = 0
df_mretts['GC'].at['Tricky Allsorts', 'MRETT21'] = 133

df_mretts['GC'] """


" df_mretts['GC'].at['eCKD 1', 'MRETT21'] = 200\ndf_mretts['GC'].at['Rasio Racing', 'MRETT21'] = 190\ndf_mretts['GC'].at['The Pedalers', 'MRETT21'] = 170\ndf_mretts['GC'].at['Pocomotion', 'MRETT21'] = 175\ndf_mretts['GC'].at['OTR Black', 'MRETT21'] = 165\ndf_mretts['GC'].at['KISS Racing Team 1', 'MRETT21'] = 150\ndf_mretts['GC'].at['Moon Riders', 'MRETT21'] = 160\ndf_mretts['GC'].at['OTR Blue', 'MRETT21'] = 155\ndf_mretts['GC'].at['OTR Green', 'MRETT21'] = 141\ndf_mretts['GC'].at['Team Lou Racing Squad', 'MRETT21'] = 137\ndf_mretts['GC'].at['eCKD 2', 'MRETT21'] = 145\ndf_mretts['GC'].at['Westerley CC Purple', 'MRETT21'] = 129\ndf_mretts['GC'].at['WKG Renegades', 'MRETT21'] = 125\ndf_mretts['GC'].at['Westerlies', 'MRETT21'] = 0\ndf_mretts['GC'].at['RGT France 1', 'MRETT21'] = 189\ndf_mretts['GC'].at['Team Lou Too', 'MRETT21'] = 0\ndf_mretts['GC'].at['Les Watts', 'MRETT21'] = 0\ndf_mretts['GC'].at['Tricky Allsorts', 'MRETT21'] = 133\n\ndf_mretts['GC'] "

	#	Total	MRETT21	MRETT22	MRETT23	MRETT24	MRETT25	MRETT26
Team								
eCKD 1	1	1000	200.0	200.0	200.0	200.0	200.0	NaN
Rasio Racing	2	920	190.0	180.0	180.0	180.0	190.0	NaN
The Pedalers	2	920	170.0	190.0	190.0	190.0	180.0	NaN
Pocomotion	4	850	175.0	165.0	170.0	165.0	175.0	NaN
OTR Black	5	835	165.0	160.0	175.0	175.0	160.0	NaN
KISS Racing Team 1	6	815	150.0	175.0	155.0	170.0	165.0	NaN
Moon Riders	7	791	160.0	141.0	160.0	160.0	170.0	NaN
OTR Blue	8	765	155.0	170.0	145.0	145.0	150.0	NaN
OTR Green	9	748	141.0	137.0	165.0	150.0	155.0	NaN
Team Lou Racing Squad	10	732	137.0	145.0	150.0	155.0	145.0	NaN
eCKD 2	11	693	145.0	133.0	137.0	137.0	141.0	NaN
Westerley CC Purple	12	669	129.0	129.0	133.0	141.0	137.0	NaN
WKG Renegades	13	520	125.0	125.0	141.0	129.0	NaN	NaN
Westerlies	14	504	NaN	121.0	129.0	125.0	129.0	NaN
RGT France 1	15	335	180.0	155.0	NaN	NaN	NaN	NaN
Team Lou Too	16	266	NaN	NaN	NaN	133.0	133.0	NaN
Les Watts	17	150	NaN	150.0	NaN	NaN	NaN	NaN
Tricky Allsorts	18	133	133.0	NaN	NaN	NaN	NaN	NaN


In [30]:
import pandas_bokeh
import bokeh
from bokeh.themes import built_in_themes
from bokeh.io import curdoc

pandas_bokeh.output_notebook()


Loading BokehJS ...

In [31]:

df_mretts['GC'][::-1].loc[:,'MRETT21':'MRETT26'].fillna(0).plot_bokeh.barh(
    ylabel="Team",
    xlabel="Points", 
    title="MRETT Series 2 GC", 
    stacked=True,
    alpha=0.6,
    figsize=(1600, 600),
    legend = "bottom_right",
    sizing_mode="scale_width"
    )

Figure(id='3913', ...)

## Convert Time column to timedelta

In [32]:
for mrett in mrett_gids.keys():
    if mrett != 'GC':
        df_mretts[mrett]['timedelta'] = pd.to_timedelta(df_mretts[mrett].Time) 
        df_mretts[mrett]['Time in Seconds'] = df_mretts[mrett]['timedelta'].dt.total_seconds() 
        df_mretts[mrett]['Time in Minutes'] = df_mretts[mrett]['Time in Seconds']/60
        df_mretts[mrett]['Time in Minutes Truncated'] = divmod(df_mretts[mrett]['Time in Seconds'], 60)[0] 
        df_mretts[mrett]['Time Remainder in Seconds'] = divmod(df_mretts[mrett]['Time in Seconds'], 60)[1] 
        df_mretts[mrett]['Team Name'] = df_mretts[mrett].index
        df_mretts[mrett]
df_mretts['21']

#        Time         Diff  Points  \
Team                                                         
eCKD 1                  1  00:33:47.1          NaN     200   
Rasio Racing            2  00:33:52.7  + 0:00:05.6     190   
RGT France 1            3  00:34:50.0  + 0:01:02.9     180   
Pocomotion              4  00:35:02.8  + 0:01:15.7     175   
The Pedalers            5  00:35:05.2  + 0:01:18.1     170   
OTR Black               6  00:35:18.2  + 0:01:31.1     165   
Moon Riders             7  00:35:38.2  + 0:01:51.1     160   
OTR Blue                8  00:35:57.6  + 0:02:10.5     155   
KISS Racing Team 1      9  00:36:00.8  + 0:02:13.7     150   
eCKD 2                 10  00:36:03.9  + 0:02:16.8     145   
OTR Green              11  00:36:08.0  + 0:02:20.9     141   
Team Lou Racing Squad  12  00:37:16.5  + 0:03:29.4     137   
Tricky Allsorts        13  00:37:44.0  + 0:03:56.9     133   
Westerley CC Purple    14  00:37:46.2  + 0:03:59.1     129   
WKG Renegades          15  00:38:38.7  + 0:04:51.6     125   

                                   timedelta  Time in Seconds  \
Team                                                            
eCKD 1                0 days 00:33:47.100000           2027.1   
Rasio Racing          0 days 00:33:52.700000           2032.7   
RGT France 1                 0 days 00:34:50           2090.0   
Pocomotion            0 days 00:35:02.800000           2102.8   
The Pedalers          0 days 00:35:05.200000           2105.2   
OTR Black             0 days 00:35:18.200000           2118.2   
Moon Riders           0 days 00:35:38.200000           2138.2   
OTR Blue              0 days 00:35:57.600000           2157.6   
KISS Racing Team 1    0 days 00:36:00.800000           2160.8   
eCKD 2                0 days 00:36:03.900000           2163.9   
OTR Green                    0 days 00:36:08           2168.0   
Team Lou Racing Squad 0 days 00:37:16.500000           2236.5   
Tricky Allsorts              0 days 00:37:44           2264.0   
Westerley CC Purple   0 days 00:37:46.200000           2266.2   
WKG Renegades         0 days 00:38:38.700000           2318.7   

                       Time in Minutes  Time in Minutes Truncated  \
Team                                                                
eCKD 1                       33.785000                       33.0   
Rasio Racing                 33.878333                       33.0   
RGT France 1                 34.833333                       34.0   
Pocomotion                   35.046667                       35.0   
The Pedalers                 35.086667                       35.0   
OTR Black                    35.303333                       35.0   
Moon Riders                  35.636667                       35.0   
OTR Blue                     35.960000                       35.0   
KISS Racing Team 1           36.013333                       36.0   
eCKD 2                       36.065000                       36.0   
OTR Green                    36.133333                       36.0   
Team Lou Racing Squad        37.275000                       37.0   
Tricky Allsorts              37.733333                       37.0   
Westerley CC Purple          37.770000                       37.0   
WKG Renegades                38.645000                       38.0   

                       Time Remainder in Seconds              Team Name  
Team                                                                     
eCKD 1                                      47.1                 eCKD 1  
Rasio Racing                                52.7           Rasio Racing  
RGT France 1                                50.0           RGT France 1  
Pocomotion                                   2.8             Pocomotion  
The Pedalers                                 5.2           The Pedalers  
OTR Black                                   18.2              OTR Black  
Moon Riders                                 38.2            Moon Riders  
OTR Blue               

In [33]:
df_mretts['21'][::-1].plot_bokeh.barh(    
    title = 'MRETT' + '21',
    xlabel = 'Time (Minutes)',
    y='Time in Minutes',
    alpha=0.6,
    figsize=(1600, 300),
    legend = "top_right",
    #hovertool_columns='Time',,
    hovertool_string="""<h2> #@{#}                             
                        @{Team Name} </h2>
                        <h3> Time: @{Time} </h3>""",
    sizing_mode="scale_width")

Figure(id='4142', ...)

In [34]:
for mrett in mrett_gids.keys():
    if mrett != 'GC':
        df_mretts[mrett][::-1].plot_bokeh.barh(    
            title = 'MRETT' + mrett,
            xlabel = 'Time (Minutes)',
            y='Time in Minutes',
            alpha=0.6,
            figsize=(1600, 300),
            legend = "top_right",
            hovertool_string="""<h2> #@{#} 
                                @{Team Name} </h2>
                                <h3> Time: @{Time} </h3>""",
            sizing_mode="scale_width")


## Save Data in Pickle Format

Use this to skip reloading/rereading 

In [35]:
for mrett in mrett_gids.keys():
    if mrett_finished[mrett] != True:
        file_name = mrett + '.pickle'
        df_mretts[mrett].to_pickle(file_name) 